File: python/Trionic_Array_II_3640.py
class Solution:

    class State:
        Up = 1
        Down = 2
        Equal = 3

    class Store:

        def __init__(self, start = None, end = None, state = None):
            self.start, self.end, self.state = start, end, state
        
        def getPrefixSum(self, nums):
            best, tot = -10**15, 0
            for i in range(self.start, self.end + 1):
                tot += nums[i]
                if i > self.start:
                    best = max(best, tot)
            return best
        
        def getSuffixSum(self, nums):
            best, tot = -10**15, 0
            for i in range(self.end, self.start - 1, -1):
                tot += nums[i]
                if i < self.end:
                    best = max(best, tot)
            return best
        
        def getSubarraySum(self, nums):
            return sum(nums[self.start : self.end + 1])

    def maxSumTrionic(self, nums: List[int]) -> int:
        
        prev, state = 0, (self.State.Up if nums[0] < nums[1] else (self.State.Down if nums[0] == nums[1] else self.State.Equal))
        seq = []

        for i in range(1, len(nums)):
            if state == self.State.Up:
                if nums[i] < nums[i - 1]:
                    seq.append(self.Store(prev, i - 1, state))
                    prev, state = i - 1, self.State.Down
                elif nums[i] == nums[i - 1]:
                    seq.append(self.Store(prev, i - 1, state))
                    prev, state = i - 1, self.State.Equal
            elif state == self.State.Equal:
                if nums[i] < nums[i - 1]:
                    seq.append(self.Store(prev, i - 1, state))
                    prev, state = i - 1, self.State.Down
                elif nums[i] > nums[i - 1]:
                    seq.append(self.Store(prev, i - 1, state))
                    prev, state = i - 1, self.State.Up
            else:
                if nums[i] == nums[i - 1]:
                    seq.append(self.Store(prev, i - 1, state))
                    prev, state = i - 1, self.State.Equal
                elif nums[i] > nums[i - 1]:
                    seq.append(self.Store(prev, i - 1, state))
                    prev, state = i - 1, self.State.Up
        seq.append(self.Store(prev, len(nums) - 1, state))

        glbSum = -10**15
        for i in range(len(seq) - 2):
            if seq[i].state == self.State.Up and seq[i + 1].state == self.State.Down and seq[i + 2].state == self.State.Up:
                curSum = seq[i].getSuffixSum(nums) + seq[i + 1].getSubarraySum(nums) + seq[i + 2].getPrefixSum(nums) - nums[seq[i].end] - nums[seq[i + 1].end]
                glbSum = max(glbSum, curSum)
        
        return glbSum

File: cpp/Trionic_Array_II_3640.cpp
class Solution {
public:

    static constexpr long long inf = 1e15;

    enum State {Up, Down, Equal};

    struct Store {
        int start;
        int end;
        State state;
        long long getPrefixSum(vector <int>& nums) {
            long long tot = 0LL, sum = -inf;
            for (int i = start; i <= end; i++) {
                tot += nums[i];
                if (i > start) {
                    sum = max(tot, sum);
                }
            }
            return sum;
        }
        long long getSuffixSum(vector <int>& nums) {
            long long tot = 0LL, sum = -inf;
            for (int i = end; i >= start; i--) {
                tot += nums[i];
                if (end > i) {
                    sum = max(tot, sum);
                }
            }
            return sum;
        }
        long long getSubarraySum(vector <int>& nums) {
            long long sum = 0LL;
            for (int i = start; i <= end; i++) {
                sum += nums[i];
            }
            return sum;
        }
    };

    long long maxSumTrionic(vector<int>& nums) {
        int prev = 0; State state = (nums[0] < nums[1] ? Up : (nums[0] > nums[1] ? Down : Equal));
        vector <Store> seq; seq.reserve(nums.size());
        for (int i = 1; i < nums.size(); i++) {
            if (state == Up) {
                if (nums[i] == nums[i - 1]) {
                    seq.emplace_back(prev, i - 1, state);
                    prev = i - 1;
                    state = Equal;
                }
                else if (nums[i] < nums[i - 1]) {
                    seq.emplace_back(prev, i - 1, state);
                    prev = i - 1;
                    state = Down;
                }
            }
            else if (state == Equal) {
                if (nums[i] > nums[i - 1]) {
                    seq.emplace_back(prev, i - 1, state);
                    prev = i - 1;
                    state = Up;
                }
                else if (nums[i] < nums[i - 1]) {
                    seq.emplace_back(prev, i - 1, state);
                    prev = i - 1;
                    state = Down;
                }
            }
            else {
                if (nums[i] == nums[i - 1]) {
                    seq.emplace_back(prev, i - 1, state);
                    prev = i - 1;
                    state = Equal;
                }
                else if (nums[i] > nums[i - 1]) {
                    seq.emplace_back(prev, i - 1, state);
                    prev = i - 1;
                    state = Up;
                }
            }
        }
        seq.emplace_back(prev, nums.size() - 1, state);
        long long globalSum = -inf;
        for (int i = 0; i + 2 < seq.size(); i++) {
            if (seq[i].state == Up && seq[i + 1].state == Down && seq[i + 2].state == Up) {
                long long curSum = seq[i].getSuffixSum(nums) + seq[i + 1].getSubarraySum(nums) + seq[i + 2].getPrefixSum(nums) - nums[seq[i].end] * 1LL - nums[seq[i + 1].end] * 1LL;
                globalSum = max(globalSum, curSum);
            }
        }
        return globalSum;
    }
};
